rfcs/convergence/array-slots.md

```diff
@@ -0,0 +1,165 @@
+    })}
+  </slots.root>
+);
+```
+
+within `useCombobox`:
+
+```js
+const state: ComboboxState = {
+  role: 'combobox',
+  tabIndex: 0,
+  ...etc,
+  components: {
+    root: 'div',
+    options: 'div',
+  },
+  options: resolveShorthandArray(props.options, {
+    role: 'option',
+    ...etc,
+  }),
+};
+```
+
+Then `resolveShorthandArray` would be exported from `resolveShorthand.ts` as something like this:
+
+```js
+export function resolveShorthandArray<Props extends Record<string, any>>(
+  value: ShorthandProps<Props>[],
+  defaultProps?: Props,
+): ObjectShorthandProps<Props>[] {
+  if (value === null || !Array.isArray(value)) {
+    return [] as ObjectShorthandProps<Props>[];
+  }
+
+  return value.map((singleValue) => resolveShorthand(singleValue, defaultProps));
+}
+```
+
+`getSlots` would be modified to handle array slots by adding a check to see if the slot is an array:
+
+```js
+if (Array.isArray(slot) {
+  slots[name] = slot.map((subSlot) => {
+    // run through the same logic getSlots currently uses to standardize slot data
+  })
+}
+```
+
+One downside to this solution is it that in order for an author to define a set of props that appear on every option, they'll need to duplicate that prop object for each option in `props.options`. We could potentially consider adding another prop like `defaultOption: React.HTMLAttributes<HTMLElement>` to make that easier.
+
+## Discarded Solutions
+
+Another option I looked into was keeping the slot itself singular, and rendering it inside a loop:
+
+In `renderCombobox`:
+
+```js
+const { slots, slotProps } = getSlots < ComboboxSlots > (state, ['option']);
+// resolveOptionProps is a way to add option-specific props like aria-selected or id based on the index
+const { optionData, resolveOptionProps } = state;
+
+return (
+  <slots.root {...slotProps.root}>
+    {optionData.map((option, i) => {
+      return <slots.option key={`option-${i}`} {...slotProps.option} {...resolveOptionProps(option, i)} />;
+    })}
+  </slots.root>
+);
+```
+
+Then `useCombobox` handles `option` and `resolveOptionProps` something like this:
+
+```js
+const state: ComboboxState = {
+  role: 'combobox',
+  tabIndex: 0,
+  ...etc,
+  components: {
+    root: 'div',
+    options: 'div',
+  },
+  options: resolveShorthand(props.option, {
+    role: 'option',
+    ...etc,
+  }),
+};
+
+const resolveOptionProps = React.useCallback(
+  (option: string | Object, index: number) => {
+    return resolveShorthand(props.option, {
+      id: `${id}-${index}`,
+      'aria-selected': index === activeIndex,
+      //simplified code, but this allows a custom per-option render:
+      children: props.renderOption ? props.renderOption(option) : option,
+    });
+  },
+  [activeIndex],
+);
+state.resolveOptionProps = resolveOptionProps;
+```
+
+The authoring experience would look like this:
+
+```js
+<Combobox option={{ 'some-default-prop-on-all-options': 'test' }} data={['apple', 'banana', 'cauliflower']}>
```

Comment:
Having been a web-dev using OUIF, this kind of shorthand API was extremely unfriendly to deal with especially when you need to support features like:
* Async loading
* Special actions rerendering a specific item
* Slightly altering specific items but not all

OUIF was very well known to have a lot of `onRenderX` callbacks or specific `onXLifecycle` methods to support this kind of shorthand API, which not only resulted in more code, but actually required mappings from our own application state to specific component lifecycle methods